DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/traffic/ixnetwork/ixnetwork_ixia_client_impl.py

```diff
@@ -108,83 +108,130 @@ def run_connect(self, device, command, *argv, **kwarg):
+
         except Exception as e:
             device.applog.info(e)
         return 0, 'Connected!'
 
+    @staticmethod
+    def __add_endpoint(dev_groups, device, topo, name, instance, is_lag=False):
+        """
+        Creates and stores L2/L3 endpoints. Works both with regular ports and lags.
+        """
+        for dev in dev_groups[name]:
+            device.applog.info('Adding device {}'.format(dev))
+            l2_name = name if is_lag else instance[1]
+            is_ipv6 = dev.get('version', 'ipv4') == 'ipv6'
+            dev_group = topo.DeviceGroup.add(Multiplier=dev.get('count', 2))
+            # L2
+            if dev.get('vlan'):
+                eth = dev_group.Ethernet.add(Name=l2_name, UseVlans=True, VlanCount=1)
+                eth.Vlan.find()[0].VlanId.Single(dev['vlan'])
+            else:
+                eth = dev_group.Ethernet.add(Name=l2_name)
+            # L3
+            if is_ipv6:
+                ep = eth.Ipv6.add(Name=dev['name'])
+                ep.Address.Increment(dev['ip'], '::1')
+            else:
+                ep = eth.Ipv4.add(Name=dev['name'])
+                ep.Address.Increment(dev['ip'], '0.0.0.1')
+            ep.GatewayIp.Single(dev['gw'])
+            ep.Prefix.Single(dev['plen'])
+
+            if dev.get('bgp_peer', {}):
+                bp = dev['bgp_peer']
+                if is_ipv6:
+                    bgp_ep = ep.BgpIpv6Peer.add(Name=dev['name'])
+                else:
+                    bgp_ep = ep.BgpIpv4Peer.add(Name=dev['name'])
+                bgp_ep.DutIp.Single(dev['gw'])
+                bgp_ep.Type.Single('external')
+                bgp_ep.LocalAs2Bytes.Single(bp['local_as'])
+                bgp_ep.HoldTimer.Single(bp['hold_timer'])
+                bgp_ep.UpdateInterval.Single(bp['update_interval'])
+                ng = dev_group.NetworkGroup.add(
+                    Multiplier=len(bp['route_ranges']), Name=dev['name']
+                )
+                ng.Enabled.Single(True)
+                for rr in bp['route_ranges']:
+                    if is_ipv6:
+                        pool = ng.Ipv6PrefixPools.add(
+                            Name=dev['name'], NumberOfAddresses=rr['number_of_routes']
+                        )
+                    else:
+                        pool = ng.Ipv4PrefixPools.add(
+                            Name=dev['name'], NumberOfAddresses=rr['number_of_routes']
+                        )
+                    pool.NetworkAddress.Single(rr['first_route'])
+                    IxnetworkIxiaClientImpl.rr_eps.append(pool)
+
+                IxnetworkIxiaClientImpl.bgp_eps.append(bgp_ep)
+            else:
+                IxnetworkIxiaClientImpl.bgp_eps.append(ep)
+                IxnetworkIxiaClientImpl.rr_eps.append(ep)
+            IxnetworkIxiaClientImpl.ip_eps.append(ep)
+            IxnetworkIxiaClientImpl.eth_eps.append(eth)
+            if is_lag:
+                IxnetworkIxiaClientImpl.raw_eps.append(instance)
```

Comment:
@SerhiyBoikoPLV  
`instance.href` instead of `instance`

Reply:
Why so? You didn't pass `lag.href` here. You were passing whatever `ixnet.Lag.add()` returned. `instance` here is taken from the `lags` dict (see L142), and the `lags` dict is created at L113